### ISSUE TITLE
support pixelfed

### ISSUE THREAD

**Issue by snarfed** [snarfed/bridgy-fed#39]
it does AP, so it might already work! first thing to do is try testing it.

https://pixelfed.org , https://github.com/pixelfed/pixelfed , cc @swentel @dansup.

alternatively we could just lobby for microformats and webmentions. 😄

**Comment by swentel** [comment]
Yes. I saw it has atom feeds, so that's great for following - would be cooler if it had microformats indeed.

I'm going to open an account and see if I can interact with it already ;-)

**Comment by swentel** [comment]
Tried it, bummer ;-)
I have a post here: https://pixelfed.social/p/swentel/16606 
Tried a reply but got stuck on the inbox post request.

Might be because federation hasn't been fully deployed yet when checking their account (https://mastodon.social/@pixelfed/100925207518316311)

**Comment by dansup** [comment]
@snarfed AP support isn't quite ready yet. I am interested in implementing more indieweb stuff like microformats and maybe webmentions!

**Comment by swentel** [comment]
Tried it from command line too, but getting 500 as well there. Will have to ping @dansup maybe in IRC to see if can do some live debugging :)

**Comment by swentel** [comment]
Been able to set it up locally, and it looks like a signature verification problem, but not 100% sure.

**Comment by swentel** [comment]
So got a bit further already setting up a pixelfed instance on my local machine. When trying to validate the actor, I was sending 'https://fed.brid.gy/r/http://realize.be' but that didn't resolve, using https://fed.brid.gy/realize.be was fine (I didn't have any problems with that testing with mastodon, and bridgy fed sends a full actor object afaics, so that might actually be a problem too).

However, after that, we need some more properties in the actor activity json as pixelfed wants to store following properties on the actor object

- summary
- outbox
- publicKey.Id

Once I faked that, the profile was saved. The username becomes @realize.be@fed.brid.gy on pixelfed, but I guess that's fine, haven't figured out exactly whether that is a problem or not.

The reply I was trying to send to a local post didn't show up though yet, trying to figure that out now.

**Comment by snarfed** [comment]
awesome progress, thank you @swentel! i can definitely add those three properties, and look into why conneg for `application/ld+json` for the wrapped reply URL isn't working.

**Comment by snarfed** [comment]
ok @swentel, bridgy fed actors now have `summary`, `outbox`, and `publicKey.id`. example: https://fed.brid.gy/realize.be .

**Comment by snarfed** [comment]
and re:

> The reply doesn't get through, because when the url is verified, it tries to get the application/ld+json representation of it

you mean that pixelfed tries to fetch a bridgy fed wrapped post URL, eg https://fed.brid.gy/r/https://snarfed.org/2018-11-10_30370 , with something like `Accept: application/ld+json`, which we need to handle and return AS2 instead of the current 302 redirect, right? just checking before i start coding.

**Comment by swentel** [comment]
Yes, the code for the internal http client uses following headers:

```
public static function zttpUserAgent()
{               
    return [
        'Accept'     => 'application/ld+json; profile="https://www.w3.org/ns/activitystreams"',
        'User-Agent' => 'PixelFedBot - https://pixelfed.org',
     ];
}
```

I'm still a bit in the dark why it does that validation again, but I assume that it wouldn't hurt to have this representation anyway. Mastodon does it as well, e.g.

```
wget --header="Accept: application/ld+json" https://mastodon.technology/@snarfed/100119606571241751
```

**Comment by snarfed** [comment]
@swentel done! `/r/...` URLs now convert to AS2 when asked for it instead of redirecting. example:

```sh
curl -v -H 'Accept: application/ld+json; profile="https://www.w3.org/ns/activitystreams"' \
  https://fed.brid.gy/r/https://snarfed.org/2018-11-10_30370
```

**Comment by swentel** [comment]
So pixelfed assumes attributedTo is a string and not an object.
It also makes assumptions on the object received from the AS2 response, and that can never work like we are sending them. So I think replies can never work from other federated platforms, maybe currently only from pixelfed instances.

This is on hold until that's figured out - and fixed in pixelfed imo.

**Comment by swentel** [comment]
Ok, I got a reply working for pixelfed. You can see the reply on https://pixelfed.social/p/swentel/16606.

This is the object I send.

```
Array
(
    [type] => Create
    [id] => https://realize.be/activitypub/104
    [actor] => https://realize.be/user/1/activitypub/swentel
    [to] => Array
        (
            [0] => https://www.w3.org/ns/activitystreams#Public
            [1] => https://pixelfed.social/users/swentel
        )

    [object] => Array
        (
            [type] => Note
            [id] => https://realize.be/reply/content/1696
            [attributedTo] => https://realize.be/user/1/activitypub/swentel
            [published] => 2018-12-25T12:09:23Z
            [content] => <p>Wow, what a view!</p>
            [inReplyTo] => https://pixelfed.social/p/swentel/16606
            [to] => Array
                (
                    [0] => https://www.w3.org/ns/activitystreams#Public
                    [1] => https://pixelfed.social/users/swentel
                )

        )

    [@context] => https://www.w3.org/ns/activitystreams
)
```

Pixelfed validates again by fetching the activitystreams representation of the id in the object (https://realize.be/reply/content/1696). The part which took me long to figure out is that the response of this is not the activity, but **only** the object

```
{"type":"Note","id":"https:\/\/realize.be\/reply\/content\/1696","attributedTo":"https:\/\/realize.be\/user\/1\/activitypub\/swentel","published":"2018-12-25T13:09:23Z","content":"\u003Cp\u003EWow, what a view!\u003C\/p\u003E","inReplyTo":"https:\/\/pixelfed.social\/p\/swentel\/16606","to":["https:\/\/www.w3.org\/ns\/activitystreams#Public","https:\/\/pixelfed.social\/users\/swentel"],"@context":"https:\/\/www.w3.org\/ns\/activitystreams"}
```

I'll come back to confirm that a note with attachment works.

**Comment by swentel** [comment]
Posting individual notes (photos) works as well. Notes without attachment won't work (unless it's a reply).
The important part here is that the attachment needs a 'Type' and 'mediaType' (mime) property. If these are not there, the validation won't work, and the post will not show up. Default valid media types (for type=image) are image/jpeg, image/png and image/gif

Payload example - visible at https://pixelfed.social/i/web/post/_/218345390364299264/219416742496178176 (but you need to be authenticated to view)

```
Array
(
    [type] => Create
    [id] => https://realize.be/activitypub/192
    [actor] => https://realize.be/user/1/activitypub/swentel
    [to] => Array
        (
            [0] => https://www.w3.org/ns/activitystreams#Public
            [1] => https://pixelfed.social/users/swentel
        )

    [object] => Array
        (
            [type] => Note
            [id] => https://realize.be/image/dartmoor
            [attributedTo] => https://realize.be/user/1/activitypub/swentel
            [published] => 2020-03-12T08:27:03Z
            [content] => <p>A magical sunrise with Vixen Tor in the background. Just haven't seen the pixies yet in the forest :)</p>
            [attachment] => Array
                (
                    [0] => stdClass Object
                        (
                            [type] => Image
                            [mediaType] => image/jpeg
                            [url] => https://realize.be/sites/default/files/DJI_0034.JPG
                        )

                )

            [to] => Array
                (
                    [0] => https://www.w3.org/ns/activitystreams#Public
                    [1] => https://pixelfed.social/users/swentel
                )

        )

    [@context] => https://www.w3.org/ns/activitystreams
)
```

**Comment by jk-na** [comment]
I tried to reply to a post on my pixelfed instance using bridgy fed and it looks like the job is failing due to the keyId being the incorrect format.

The error at the pixelfed end is: `ErrorException: Undefined index: keyId in /usr/share/nginx/html/pixelfed/app/Jobs/InboxPipeline/InboxValidator.php:159`

The format of the keyId in the post is: `keyId="acct:jk.nipponalba.scot@jk.nipponalba.scot"`
Whereas on a randomly selected successful job it is in the format: `keyId="https://pipou.academy/users/AuroreAcelya#main-key"`

**Comment by snarfed** [comment]
I tried to federate [a like](https://snarfed.org/2023-01-17_piconic) of [this post](https://piconic.co/i/web/post/507466431055899254) on piconic.co, running Pixelfed v0.11.4. It succeeded, [BF log here](https://fed.brid.gy/log?start_time=1673983311&key=https://snarfed.org/2023-01-17_piconic%20https://piconic.co/p/Ovadex/507466431055899254&module=), but I don't see the like on the post yet. 🤷

**Comment by snarfed** [comment]
I liked the same post from @snarfed@indieweb.social (Mastodon), and that like now shows up on the post, so interop here still isn't quite working yet.

**Comment by snarfed** [comment]
Another thing to look at: when I search for @snarfed.org@snarfed.org on a PixelFed instance, it finds me, but it shows my handle as @snarfed.org@fed.brid.gy. 😐

**Comment by snarfed** [comment]
After #442, I tried liking and replying to the same piconic.co post above again, https://piconic.co/i/web/post/507466431055899254 , both 200ed but neither showed up.

**Comment by snarfed** [comment]
Tried this follow just now, also 200ed and didn't show up.

If BF users' server shows up in PixelFed as fed.brid.gy, not their domain, ie we're eg `@snarfed.org@fed.brid.gy`, I wonder if that's related. Maybe a signature `keyId` thing? Should look more.

```json
{
  "published": "2023-10-24T13:01:58-07:00",
  "content": "Follows <a class=\"u-follow-of\" href=\"https://piconic.co/i/web/profile/145384761425268736\">https://piconic.co/i/web/profile/145384761425268736</a>.\n<div class=\"e-content\">\npiconic\n</div>",
  "url": "https://fed.brid.gy/r/https://snarfed.org/2023-10-24_piconic-2",
  "actor": "https://fed.brid.gy/snarfed.org",
  "id": "https://fed.brid.gy/r/https://snarfed.org/2023-10-24_piconic-2",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Follow",
  "object": "https://piconic.co/users/snarfed",
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ],
  "contentMap": {
    "en": "Follows <a class=\"u-follow-of\" href=\"https://piconic.co/i/web/profile/145384761425268736\">https://piconic.co/i/web/profile/145384761425268736</a>.\n<div class=\"e-content\">\npiconic\n</div>"
  }
}
```

**Comment by snarfed** [comment]
Follows still aren't getting accepted ^. Eg for @TomCasavant in #1053, we sent the `Follow` below from actor `https://bsky.brid.gy/bsky.brid.gy` to https://pixelfed.social/users/mrpresidenttom/inbox at 2024-05-08T10:45:02.854174+00:00, and it returned HTTP 202, but never sent back an `Accept`, and the follow never registered on pixelfed.social's end.

```json
{
  "id": "https://bsky.brid.gy/r/https://bsky.brid.gy/#follow-back-https://pixelfed.social/users/mrpresidenttom-2024-05-08T10:45:02.854174+00:00",
  "actor": "https://bsky.brid.gy/bsky.brid.gy",
  "@context": "https://www.w3.org/ns/activitystreams",
  "type": "Follow",
  "object": "https://pixelfed.social/users/mrpresidenttom",
  "to": [
    "https://www.w3.org/ns/activitystreams#Public"
  ]
}
```

**Comment by snarfed** [comment]
Pinged dansup on Matrix just now 🤞

**Comment by ThisIsMissEm** [comment]
@snarfed that object ID isn't a validate URI as far as I know? Maybe that's the source of the issue?

**Comment by snarfed** [comment]
@ThisIsMissEm thanks for looking! And hmm. `https://bsky.brid.gy/r/https://bsky.brid.gy/#follow-back-https://pixelfed.social/users/mrpresidenttom-2024-05-08T10:45:02.854174+00:00` isn't a valid URI? Because of the non-encoded `+` or `:`s? Lots of other fediverse software is accepting that id format fine, so I'm guessing it's not the issue here, but I'd be curious to hear more if it actually is invalid somehow.

**Comment by qazmlp** [comment]
> @ThisIsMissEm thanks for looking! And hmm. `https://bsky.brid.gy/r/https://bsky.brid.gy/#follow-back-https://pixelfed.social/users/mrpresidenttom-2024-05-08T10:45:02.854174+00:00` isn't a valid URI? Because of the non-encoded `+` or `:`s? Lots of other fediverse software is accepting that id format fine, so I'm guessing it's not the issue here, but I'd be curious to hear more if it actually is invalid somehow.

The URI is syntactically fine afaict, but maybe it's trying to actually fetch it?

**Comment by snarfed** [comment]
Hmm, maybe? ids with fragment are relatively common in the fediverse, and how to fetch them is largely undefined. 🤷 https://github.com/w3c/activitypub/issues/367

**Comment by snarfed** [comment]
> Pinged dansup on Matrix just now 🤞
No answer, as always 😕

**Comment by snarfed** [comment]
Interestingly bridging _out_of Pixelfed into Bluesky is working ok, at least somewhat, eg https://bsky.app/profile/freyja-lynx.pixelfed.social.ap.brid.gy . That bridged account doesn't have the latest post from the Pixelfed account though, not sure why.

**Comment by lucidillusions** [comment]
Adding some of my observations 
1. Created a new pixelfed.social account
2. 15 days later, followed bridgy account
3. Updated my handle to look pretty on the bsky side
4. Updated insta archive on pixelfed, (no notification on bsky or mastodon)
5. Uploaded new post on pixelfed - the notifications appears on mastodon (obviously) - but nothing on bsky.
6. Looked around bridged account, and saw 3-4 accounts that had posted a day or two before on pixelfed, and those appeared on their bridged account. 
7. Had a doubt, went and deleted the TXT file from my domain records - updated pixelfed account from fed.brid.gy
8. Uploaded new post on pixelfed > noticed the post appeared on threads
9. Went to bsky, the post had bridged itself (except the handle was invalid for time being)
10. Sent a DM to bridgy account on pixelfed to update handle back to *.brid.gy  and updated profile.
11. Handle shows fine on bsky. 
12. Uploaded new photos on pixelfed, and the posts bridge almost instantly to bsky. 

Key observations 
1. updating the bsky handle breaks the bridge somehow
2. threads was somehow affected by the above step (cannot make sense of the logic at all) 
3. comments on bsky doesn't bridge to pixelfed
4. likes/repost shows up across bridge. 
5. Alt text bridges fine from pixelfed to bsky 

the bridged account: https://bsky.app/profile/lucidillusions.pixelfed.social.ap.brid.gy

**Comment by lucajet** [comment]
I had tried some months ago, and since nothing was passed, i deleted the bridged account by blocking the bridge.
When i saw someone succeeding i unblocked the bridge and the account bs side war recreated but it took a dm-conversation via dm (i said start and then did) to be followed back again.

I confirm point 3 being the main issue, if someone comment on bsky, with a bridged account, that comment does not reach back pixelfed.

Example, this picture

https://pixelfed.uno/i/web/post/783314454576926845 is last post on 
https://fed.brid.gy/ap/@underscorner@pixelfed.uno ( don't know why there are no logs)

wash shared,liked and commented from a bridged bsky users but the comment didn't reached back pixelfed so i can't answer

![immagine](https://github.com/user-attachments/assets/11816ff5-363a-4ec5-8578-80a9cffa1133)

**Comment by snarfed** [comment]
Thanks for the example @lucajet! Looking at https://fed.brid.gy/ap/@underscorner@pixelfed.uno/notifications , BF is seeing the replies, likes, and reposts on Bluesky and sending them to pixelfed.uno as AP activities. As an example, for the reply https://bsky.app/profile/did:plc:vdmjwle5fbxhisu5qdf5kc7g/post/3lfhruoafek2v , BF POSTed the activity below to https://pixelfed.uno/users/underscorner/inbox at 2025-01-11 05:12:07 and got an HTTP 200 response. cc @dansup...?

```json
{
  "@context": ["https://www.w3.org/ns/activitystreams"],
  "type": "Create",
  "id": "https://bsky.brid.gy/convert/ap/at://did:plc:vdmjwle5fbxhisu5qdf5kc7g/app.bsky.feed.post/3lfhruoafek2v#bridgy-fed-create",
  "actor": "https://bsky.brid.gy/ap/did:plc:vdmjwle5fbxhisu5qdf5kc7g",
  "published": "2025-01-11T13:12:07.206672+00:00",
  "object": {
    "type": "Note",
    "id": "https://bsky.brid.gy/convert/ap/at://did:plc:vdmjwle5fbxhisu5qdf5kc7g/app.bsky.feed.post/3lfhruoafek2v",
    "url": "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:vdmjwle5fbxhisu5qdf5kc7g/post/3lfhruoafek2v",
    "attributedTo": "https://bsky.brid.gy/ap/did:plc:vdmjwle5fbxhisu5qdf5kc7g",
    "content": "<p>Che bello</p>",
    "published": "2025-01-11T13:12:06.515Z",
    "contentMap": {"it": "Che bello"},
    "inReplyTo": "https://pixelfed.uno/p/underscorner/783314454576926845",
    "cc": [
      "https://pixelfed.uno/users/underscorner",
      "https://www.w3.org/ns/activitystreams#Public",
      "https://pixelfed.uno/users/underscorner/followers"
    ],
    "tag": [{
      "type": "Mention",
      "href": "https://pixelfed.uno/users/underscorner"
    }],
    "to": ["https://www.w3.org/ns/activitystreams#Public"]
  },
  "to": ["https://www.w3.org/ns/activitystreams#Public"],
  "cc": [
    "https://pixelfed.uno/users/underscorner",
    "https://www.w3.org/ns/activitystreams#Public",
    "https://pixelfed.uno/users/underscorner/followers"
  ]
}
```

**Comment by snarfed** [comment]
Similar interop problem with pixelfedsweden.se in #1732: likes, replies, etc aren't showing up there. It's running Pixelfed  v0.12.4.

Example reply:
```
2025-01-28 14:48:59.099 PST
requests.post https://pixelfedsweden.se/users/DoktorZjivago/inbox {'data': b'{
  "@context": ["https://www.w3.org/ns/activitystreams"],
  "type": "Create",
  "id": "https://bsky.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3lgtjzqv4k22x#bridgy-fed-create",
  "actor": "https://bsky.brid.gy/ap/did:plc:3ljmtyyjqcjee2kpewgsifvb",
  "published": "2025-01-28T22:48:56.739068+00:00",
  "object": {
    "type": "Note",
    "id": "https://bsky.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.post/3lgtjzqv4k22x",
    "url": "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:3ljmtyyjqcjee2kpewgsifvb/post/3lgtjzqv4k22x",
    "attributedTo": "https://bsky.brid.gy/ap/did:plc:3ljmtyyjqcjee2kpewgsifvb",
    "inReplyTo": "https://pixelfedsweden.se/p/DoktorZjivago/789976043183698130",
    "content": "<p>\\ud83c\\udf89</p>",
    "contentMap": {"en": "\\ud83c\\udf89"},
    "published": "2025-01-28T22:48:55.617Z",
    "tag": [{
        "type": "Mention",
        "href": "https://pixelfedsweden.se/users/DoktorZjivago"
      }],
    "to": ["https://www.w3.org/ns/activitystreams#Public"],
    "cc": [
      "https://pixelfedsweden.se/users/DoktorZjivago",
      "https://www.w3.org/ns/activitystreams#Public",
      "https://pixelfedsweden.se/users/DoktorZjivago/followers"
    ]
  },
  "to": ["https://www.w3.org/ns/activitystreams#Public"],
  "cc": [
    "https://pixelfedsweden.se/users/DoktorZjivago",
    "https://www.w3.org/ns/activitystreams#Public",
    "https://pixelfedsweden.se/users/DoktorZjivago/followers"
  ]
}'}
```

Example like:
```
2025-01-28 14:48:49.339 PST
requests.post https://pixelfedsweden.se/users/DoktorZjivago/inbox {'data': b'{
  "@context": ["https://www.w3.org/ns/activitystreams"],
  "type": "Like",
  "id": "https://bsky.brid.gy/convert/ap/at://did:plc:3ljmtyyjqcjee2kpewgsifvb/app.bsky.feed.like/3lgtjzkaa6e2e",
  "actor": "https://bsky.brid.gy/ap/did:plc:3ljmtyyjqcjee2kpewgsifvb",
  "url": "https://fed.brid.gy/r/https://bsky.app/profile/did:plc:rzvef4zrevnznp37uacidwy6/post/3lgthi47uv7p2#liked_by_did:plc:3ljmtyyjqcjee2kpewgsifvb",
  "object": "https://pixelfedsweden.se/p/DoktorZjivago/789976043183698130",
  "to": ["https://www.w3.org/ns/activitystreams#Public"],
  "cc": [
    "https://pixelfedsweden.se/users/DoktorZjivago",
    "https://www.w3.org/ns/activitystreams#Public",
    "https://pixelfedsweden.se/users/DoktorZjivago/followers"
  ]
}'}
```

**Comment by Goldmaster** [comment]
I have been able to use pixelfed.social with bridgy fed. Simply search for bridgy fed from pixelfed then follow just like normal. Your pixelfed account would then show up on Bluesky.

**Comment by snarfed** [comment]
Thanks for the nudge @Goldmaster! I've also confirmed that likes and reposts are also flowing back and forth now. Example:  https://pixelfed.social/p/quillmatiq/785930915806587556

I'm not seeing bridged replies show up on Pixelfed though. Has anyone seen a Bluesky or web reply on a Pixelfed post yet?

**Comment by lucidillusions** [comment]
@snarfed not sure if this would help. 
1. Original post: https://pixelfed.social/p/lucidillusions/819443420804496794
2. Replied to it on masto social, which shows up on pixelfed as well as bsky (bridged): https://mastodon.social/@lucidillusions/114371143382255376
3. The above reply (bridged on bsky): https://bsky.app/profile/fedi.lucidillusions.in/post/3lnazqmowa5p2

I made a bsky comment to both the bridged posts - the reply to masto social shows up on MS - but not on PF. 
I was hoping the reply to PF post would show up on Masto at least, but it doesn't.

**Comment by h-2** [comment]
Hey! Just checking in here, interested in what the current state is.

I am following some bridged bsky-accounts from my pixelfed account, but I can't see anything they post.

Here's an example:
* Original account: https://bsky.app/profile/nergilpainting.bsky.social
* Fully accessible from Mastodon: https://mastodon.social/@nergilpainting.bsky.social@bsky.brid.gy
* But no content from Pixelfed: https://pixelfed.social/i/web/profile/765708539383482279

**Comment by h-2** [comment]
Furthermore, it says "Invalid Handle" on my bridged profile with Bsky: https://bsky.app/profile/heretic-hannes.pixelfed.social.ap.brid.gy and the profile cannot be opened from the search.

**Comment by snarfed** [comment]
@h-2 thanks for the nudge! Your Pixelfed account's bridged Bluesky handle looks ok now, may just have been a brief glitch.

As for why bridged Bluesky posts aren't visible, seems like we're in the same position as when we looked at this last over a year ago, https://github.com/snarfed/bridgy-fed/issues/39#issuecomment-2817226103 . I think I've done as much investigation as I can on Bridgy Fed's end, at this point we need a Pixelfed developer and/or admin to look on their end.
